invokeai/app/services/config.py

```diff
@@ -0,0 +1,365 @@
+resolve paths relative to the runtime root directory. They each return
+a Path object:
+
+ root_path          - path to InvokeAI root
+ output_path        - path to default outputs directory
+ model_conf_path    - path to models.yaml
+ conf               - alias for the above
+ embedding_path     - path to the embeddings directory
+ lora_path          - path to the LoRA directory
+ 
+
+'''
+import argparse
+import os
+import sys
+from argparse import ArgumentParser
+from omegaconf import OmegaConf, DictConfig
+from pathlib import Path
+from pydantic import BaseSettings, Field, parse_obj_as
+from typing import Any, ClassVar, List, Literal, Union, get_origin, get_type_hints, get_args
+
+INIT_FILE = Path('invokeai.yaml')
+LEGACY_INIT_FILE = Path('invokeai.init')
+
+class InvokeAISettings(BaseSettings):
+    '''
+    Runtime configuration settings in which default values are
+    read from an omegaconf .yaml file.
+    '''
+    initconf             : ClassVar[DictConfig] = None
+
+    def parse_args(self, argv: list=sys.argv[1:]):
+        parser = self.get_parser()
+        opt, _ = parser.parse_known_args(argv)
+        for name in self.__fields__:
+            if name not in self._excluded():
+                setattr(self, name, getattr(opt,name))
+
+    @classmethod
+    def add_parser_arguments(cls, parser):
+        env_prefix = cls.Config.env_prefix if hasattr(cls.Config,'env_prefix') else 'INVOKEAI_'
+        default_settings_stanza = get_args(get_type_hints(cls)['type'])[0]
+        initconf = cls.initconf.get(default_settings_stanza) if cls.initconf and default_settings_stanza in cls.initconf else None
+
+        fields = cls.__fields__
+        for name, field in fields.items():
+            if name not in cls._excluded():
+                env_name = env_prefix+f'{cls.cmd_name()}_{name}'
+                if initconf and name in initconf:
+                    field.default = initconf.get(name)
+                if env_name in os.environ:
+                    field.default = os.environ[env_name]
+                cls.add_field_argument(parser, name, field)
+
+
+    @classmethod
+    def cmd_name(self, command_field: str='type')->str:
+        hints = get_type_hints(self)
+        return get_args(hints[command_field])[0]
+
+    @classmethod
+    def get_parser(cls)->ArgumentParser:
+        parser = ArgumentParser(
+            prog=cls.cmd_name(),
+            description=cls.__doc__,
+        )
+        cls.add_parser_arguments(parser)
+        return parser
+
+    @classmethod
+    def add_subparser(cls, parser: argparse.ArgumentParser):
+        parser.add_parser(cls.cmd_name(), help=cls.__doc__)
+
+    @classmethod
+    def _excluded(self)->List[str]:
+        return ['type','initconf']
+    
+    class Config:
+        env_file_encoding = 'utf-8'
+        arbitrary_types_allowed = True
+        env_prefix = 'INVOKEAI_'
+        extra = 'allow'      # FIX: this shouldn't be allowed, but without it graphs fail to validate
```

Comment:
I see this was added to the LibraryGraph definition. Why do we need to allow extra fields?

Reply:
I have been struggling to understand why this is needed. Without it, pydantic returns validation errors whenever a LibraryGraph is created. This is not the case for nodes or commands. As the comment says, I'm uncomfortable with this workaround.

There is some deep voodoo going on in LibraryGraphs that I haven't figured out yet, but I'll take another crack at it before reaching out to Kyle for help.

Reply:
@Kyle0654 

I think if you inherit from BaseInvocation you shouldn't need to explicitly add the `id` field, but I'm not familiar with the cli enough.